### PR TITLE
VACMS-1389: post api

### PR DESCRIPTION
**Pull request by oksana-c**
## Description

See #1389 . 

## Testing done

Visual + CR of this PR and https://git.drupalcode.org/project/post_api/tree/8.x-1.x

## Screenshots

Queue UI - /admin/config/post-api/queue

Screen Shot 2020-04-10 at 3 59 18 PM

## QA steps

As user 1
1. Update settings.php file w/ facility endpoint url and apikey in order to test locally https://github.com/department-of-veterans-affairs/va.gov-cms/pull/1446/files#diff-2737999ec8b2368b193b40b46c5679c2R154
1.  Edit or update any facility node
- [x] visit /admin/config/post-api/queue - verify queue item is added w/ correct 
image

- [x] Process queue manually by pressing "Process queue" button
image

1. confirm your updated data is posted by performing a curl GET request for the facility ID you edited 
`curl -v -H "apikey: <APIKEY>" https://<HOST>/services/va_facilities/v0/facilities/<FACILITY_ID>`
- [x]  add/edit the same facility twice, confirm only item w/ latest data is queued (old record deduped)
1. run Cron - confirm facility queue is processed by running curl against edited/added facility ID again
1. test queue failure and Slack notification: 
    1. enter bogus apikey in settings.php
    1. update `$config['slack.settings']['slack_webhook_url'] =` w/ Slack webhook url (you'll have to use your own slack group account to test if available , or reach out to Elijah to get cam-notification webhook via keybase)
    1. and run queue worker via UI
    - [x] Confirm error message is in log and shows correct number of items left in queue.
    1. Confirm Slack notification went out - :triangular_flag_on_post: Post API has failed to process queued items. Total items in queue: %d.

## Definition of Done
- [ ] Product release notes 
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

File: docroot/modules/custom/va_gov_post_api/va_gov_post_api.module [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Implements hook_entity_update().
+ *
+ * Queues updated items for sync.
+ */
+function post_api_entity_update(EntityInterface $entity) {
+  _post_api_add_facility_to_queue($entity);
+}
+
+function _post_api_add_facility_to_queue(EntityInterface $entity) {
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:28]: needs docblock.

File: docroot/modules/custom/va_gov_post_api/va_gov_post_api.module [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    'vet_center',
+  ];
+
+  if ($entity->getEntityTypeId() === 'node' && in_array($entity->bundle(), $facility_types)) {
+    $queue = Drupal::service('post_api.add_to_queue');
+
+    $facility_id = $entity->hasField('field_facility_locator_api_id') ? $entity->field_facility_locator_api_id->value : NULL;
+
+    // Queue item's Unique ID.
+    $data['uid'] = 'facility_status_' . $facility_id;
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:42]: Do we need to instantiate this $data array first?

File: docroot/modules/custom/va_gov_post_api/va_gov_post_api.module [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    // Set a key based on which the endpoint will be
+    // defined during queue execution.
+    $data['endpoint_path'] = '/services/va_facilities/v0/facilities/' . $facility_id . '/cms-overlay';
+
+    // Set payload. Default payload provided by this module is empty. You
+    // must set your own payload in custom Payload Decorator.
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:49]: Outdated comments?

File: docroot/modules/custom/va_gov_post_api/va_gov_post_api.module (hunk 2 above)
Comment (oksana-c) [reply docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:42]: no, but this made me think of another fallback I have to add. thank you! and please stand by

File: docroot/modules/custom/va_gov_post_api/va_gov_post_api.module (hunk 1 above)
Comment (oksana-c) [reply docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:28]: can't believe phpcs missed this

File: docroot/modules/custom/va_gov_post_api/va_gov_post_api.module (hunk 3 above)
Comment (oksana-c) [reply docroot/modules/custom/va_gov_post_api/va_gov_post_api.module:49]: indeed

**Comment by swirtSJW** [comment]
Adding something to the queue, then triggering cron causes it to be removed from the queue but there is no dblog entry that slack is disabled and logged to dblog   the way there is when I trigger the queue by hand.

image

**Comment by oksana-c** [comment]
@swirtSJW , let's look at this w/ screenshare.  My initial understanding of what you described is that the behavior is expected and the slack error would not be placed in dblog if the item was indeed processed/ removed from queue.

**Comment by swirtSJW** [comment]
> @swirtSJW , let's look at this w/ screenshare. My initial understanding of what you described is that the behavior is expected and the slack error would not be placed in dblog if the item was indeed processed/ removed from queue.

Ok so I should only expect to see that notification when running it by hand.    All is good.

**Pull request merged**